### ISSUE TITLE
Truncated or oversized response headers received from daemon process

### ISSUE THREAD

**Issue by viktorkh** [GrahamDumpleton/mod_wsgi#696]
Hi
I'm running on our backend  apache2 on python3 EC2  ma5.4xlarge instance (64 Gb memory, 16 vCpu ) with Ubuntu 16.04.7 my database is on AWS RDS (r5.4xlarge)
Our backend server serve several sites (around 10 customers)
The apache2.conf settings
Timeout 900
KeepAlive On
MaxKeepAliveRequests 700
KeepAliveTimeout 2
HostnameLookups Off
AccessFileName .htaccess
<FilesMatch "^\.ht">
        Require all denied
</FilesMatch>
The mod_wsgi settings - each site load on his file config(have 10 sites)
site1.conf
WSGIDaemonProcess site1  python-path=/mnt/ttt/py/production/pythonn home=/mnt/ttt/py/production/python processes=12 threads=5 display-name=%{GROUP}
WSGIScriptAlias /site1 /mnt/ttt/py/production/python/portal/wsgi_main.py
WSGIApplicationGroup %{GLOBAL}
<Location /site1>
WSGIProcessGroup site1
</Location>

Given my configuration, could you please let me know if the maximum number of processes/threads/requests is ok?
We are experiencing performance issues in the transition to WSGI what may the reason?

Also, we got the below error from wsgi log 
 - [pid 7023:tid 139816746346240] [client 127.0.0.1:37026] Timeout when reading response headers from daemon process 'site1': /mnt/ttt/py/production/python/portal/wsgi_main.py

- [Mon Jun 21 06:25:01.449450 2021] [wsgi:error] [pid 2010:tid 139816737953536] (2)No such file or directory: [client 127.0.0.1:63706] mod_wsgi (pid=2010): Unable to connect to WSGI daemon process 'site2' on '/var/run/apache2/wsgi.26625.0.9.sock' as user with uid=1100.

- [Mon Jun 21 06:25:05.074549 2021] [wsgi:error] [pid 6925:tid 139816746346240] [client 127.0.0.1:62090] Truncated or oversized response headers received from daemon process 'site1':/mnt/ttt/py/production/python/portal/wsgi_main.py

**Comment by GrahamDumpleton** [comment]
Is that the complete Apache config or are you cherry picking select configuration?

Only ask as sometimes see people who toss out all the default Apache configuration and try and add their own minimal configuration. That is generally a bad idea.

Performance issues can be due to many things. I can't say whether your configuration is adequate because I know nothing about your application and the amount of traffic it gets.

As to the error:

```
[Mon Jun 21 06:25:01.449450 2021] [wsgi:error] [pid 2010:tid 139816737953536] (2)No such file or directory: [client 127.0.0.1:63706] mod_wsgi (pid=2010): Unable to connect to WSGI daemon process 'site2' on '/var/run/apache2/wsgi.26625.0.9.sock' as user with uid=1100.
```

with others also possibly occurring at same time, this is usually because the system logrotate service is forcing Apache to do a graceful restart, but your use of keep alive connections when that occurs can cause issues as the Apache child worker process being kept around for the keep alive connection end up with stale connection details for the mod_wsgi daemon process, which also got forced to restart.

You can avoid the error by setting the option:

```
WSGISocketRotation Off
```

See:

* https://modwsgi.readthedocs.io/en/master/release-notes/version-4.6.0.html?highlight=WSGISocketRotation#new-features

What version of mod_wsgi are you using. If you are using the latest version I can give you some code to put in the WSGI script file that can dump out some data about request throughput, capacity usage, response times etc.

**Comment by viktorkh** [comment]
Hi,
No, this is not a complete apache2 conf file, see attached zip with complete files configuration.
[conf_files.zip](https://github.com/GrahamDumpleton/mod_wsgi/files/6685981/conf_files.zip)

Also, we are working with mpm_worker_module see below the conf 

<IfModule mpm_worker_module>
        StartServers       1
        ServerLimit       50
        MaxClients        150
        MinSpareThreads   25
        MaxSpareThreads   27
        ThreadsPerChild     3
        MaxRequestsPerChild  10000
</IfModule>

My wsgi version
mod-wsgi              4.8.0

Thanks for your help

**Comment by GrahamDumpleton** [comment]
This configuration is sort of broken. ``MixSpareThread`` and ``MaxSpareThreads`` should usually be a multiple of ``ThreadsPerChild`` else can end up with strange effects in how Apache auto scales processes.

Why have you set ``ThreadsPerChild`` to 3 in the first place? What are you hoping to achieve with that configuration?

Usually for Apache worker processes you wouldn't have so few threads per child process and with your other configuration means Apache would be spinning up a log of child worker process and recycling them when not necessary causing increased load.

Since for you configuration you are delegating each WSGI application to a separate daemon process group, suggest setting:

```
WSGIRestrictEmbedded On
```

outside of any ``VirtualHost``.

This will prevent Python being initialised in Apache child worker processes, which with your configuration would just cause more load on child worker processes when starting, when Python isn't even used in those processes.

Also, instead of:

```
WSGIDaemonProcess site1  python-path=/mnt/tkd1/ttt/production/python home=/mnt/tkd1/ttt/production/python  processes=12 threads=5 display-name=%{GROUP}
WSGIScriptAlias /site1 /mnt/tkd1/ttt/production/python/portal/wsgi_main.py
WSGIApplicationGroup %{GLOBAL}
<Location /site1>
WSGIProcessGroup site1
</Location>
```

use this form for the configuration of each WSGI application.

```
WSGIDaemonProcess site1  python-path=/mnt/tkd1/ttt/production/python home=/mnt/tkd1/ttt/production/python  processes=12 threads=5 display-name=%{GROUP}
WSGIScriptAlias /site1 /mnt/tkd1/ttt/production/python/portal/wsgi_main.py process-group=site1 application-group=%{GLOBAL}
```

It is more concise and less prone to mistakes happening if don't set ``WSGIProcessGroup`` and ``WSGIApplicationGroup`` in correct context, which you are sort of doing wrong with ``WSGIApplicationGroup``.

Also worried that you have so many processes for every mod every mod_wsgi daemon process group. You seem to be setting capacity on the excessive side.

What is the request throughput rate for each site, average response times, and longest response times?

Would also suggesting setting Apache log level to info instead of error.

```
LogLevel info
```

That way you can see when mod_wsgi is restarting daemon processes and why. Something could be causing your process to churn for some reason. Eg. processes crashing.

Also suggesting watching:

* https://www.youtube.com/watch?v=k6Erh7oHvns
* https://www.youtube.com/watch?v=eOkxLCCbU9w
* https://www.youtube.com/watch?v=CPz0s1CQsTE
* https://www.youtube.com/watch?v=SGleKfigMsk

Anyway, answer questions above and can then look at suggesting code to stick in WSGI script files to log stats on request throughput, response times, capacity usage etc, which will help with better setting number of processes/threads.

**Comment by viktorkh** [comment]
Hi Graham
I want to thank you for the help and very much and appreciate the time you spend helping me solve the problems.
I implemented your recommendations and run the system again ,we see little bit improvement  still see some performance issue and errors on wsgi log.
we are in middle of   migration process from python2 with mod_python to python3 with mod_wsgi.  
There are  some  performance  problem with python3 and mod_wsgi in our backend process.
Our backend server runs time-series transformations and anomaly analysise ,these are Python algorithms called by http request to alg.py.
The clients requests come from java  process ,all requests are internally from the local host backend server.

See below the errors from wsgi log
[Wed Jun 23 23:29:56.525536 2021] [wsgi:error] [pid 20441:tid 140504570164992] [client 127.0.0.1:64742] Timeout when reading response headers from daemon process 'evides': /mnt/tkd1/takadu/production/python/portal/wsgi_main.py
[Thu Jun 24 06:25:05.859091 2021] [wsgi:error] [pid 20445:tid 140504419235584] [client 127.0.0.1:18314] Truncated or oversized response headers received from daemon process 'kub': /mnt/tkd1/takadu/production/python/portal/wsgi_main.py

Give our ec2 instance r5.4xlarge instance (128 Gb memory, 16 vCpu ) below are the new mpm_worker config
<IfModule mpm_worker_module>
        StartServers      16
        ServerLimit       16
        MinSpareThreads   256
        MaxSpareThreads   512
        ThreadsPerChild     32
        MaxRequestWorkers 512
        MaxRequestsPerChild  10000
</IfModule>
Also attached the logs ,monitor and new config files
[ttt_24_06_.zip](https://github.com/GrahamDumpleton/mod_wsgi/files/6707756/ttt_24_06_.zip)

**Comment by GrahamDumpleton** [comment]
Since there is no evidence of daemon processes crashing or restarting themselves, it would look like the request handlers in the daemon processes are deadlocking or hang on waiting for some back end service.

What I would do at this point is set a ``request-timeout`` on the mod_wsgi daemon process group. See option description in:

* https://modwsgi.readthedocs.io/en/master/configuration-directives/WSGIDaemonProcess.html#request-timeout

This will cause the daemon processes to restart automatically when you have a request which runs for a very long time.

This will help to unstick the process so you don't have to manually restart the whole of Apache.

Also possibly add ``queue-timeout`` so when recovering, any back log of requests gets thrown away so you can clear out backlog and catch up to latest requests and not keep being backlogged.

* https://modwsgi.readthedocs.io/en/master/configuration-directives/WSGIDaemonProcess.html#queue-timeout

How long you set ``request-timeout`` will depend on how long your requests are expected to run for. Since they are going to get failed anyway when the default Apache socket ``Timeout`` is exceeded, you can start by setting it the same as whatever you have that set to.

One side effect of adding this is that when the daemon process is restarted there should be an attempt to log all the Python stack traces of currently running requests. This way you can see what the threads were doing and thus where they may be getting stuck.

The alternative to getting the stack traces is a more manual method using code in:

* https://modwsgi.readthedocs.io/en/master/user-guides/debugging-techniques.html#extracting-python-stack-traces

That code is old though so you will need to update it to Python 3.

When you see Apache is stuck, then trigger the dump of the stack traces using that.

**Comment by viktorkh** [comment]
We applied your configuration.
For now we have  high CPU usage .
Our architecture is using python as a backend server which receives http requests by apache server and does some calculations. 
As you saw , we have multple sites: /site1,  /site2, ...
all of them use the same python code directory:  home=/tkd/prod/python and 
Aliases refers to the same wsgi file:
WSGIScriptAlias /site1 /mnt/python/portal/wsgi_main.py
WSGIScriptAlias /site2 /mnt/python/portal/wsgi_main.py
.....
we defined for each site: thread=3 proccess=12 

Do we need to run it on a single WSGIDaemonProcess or one per site ?

BTW, our prod runs on mod_python , there isn't problems with CPU.
Which brings me to the conclusion that our problems in configuration

Any idea?
